zcash_keys/src/keys/transparent.rs

```diff
@@ -0,0 +1,340 @@
+            let middle = [
+                0xA0, 0x81, 0xA5, 0x30, 0x81, 0xA2, 0x02, 0x01, 0x01, 0x30, 0x2C, 0x06, 0x07, 0x2A,
+                0x86, 0x48, 0xCE, 0x3D, 0x01, 0x01, 0x02, 0x21, 0x00, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+                0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+                0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE, 0xFF, 0xFF, 0xFC, 0x2F, 0x30,
+                0x06, 0x04, 0x01, 0x00, 0x04, 0x01, 0x07, 0x04, 0x41, 0x04, 0x79, 0xBE, 0x66, 0x7E,
+                0xF9, 0xDC, 0xBB, 0xAC, 0x55, 0xA0, 0x62, 0x95, 0xCE, 0x87, 0x0B, 0x07, 0x02, 0x9B,
+                0xFC, 0xDB, 0x2D, 0xCE, 0x28, 0xD9, 0x59, 0xF2, 0x81, 0x5B, 0x16, 0xF8, 0x17, 0x98,
+                0x48, 0x3A, 0xDA, 0x77, 0x26, 0xA3, 0xC4, 0x65, 0x5D, 0xA4, 0xFB, 0xFC, 0x0E, 0x11,
+                0x08, 0xA8, 0xFD, 0x17, 0xB4, 0x48, 0xA6, 0x85, 0x54, 0x19, 0x9C, 0x47, 0xD0, 0x8F,
+                0xFB, 0x10, 0xD4, 0xB8, 0x02, 0x21, 0x00, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+                0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE, 0xBA, 0xAE, 0xDC, 0xE6, 0xAF,
+                0x48, 0xA0, 0x3B, 0xBF, 0xD2, 0x5E, 0x8C, 0xD0, 0x36, 0x41, 0x41, 0x02, 0x01, 0x01,
+                0xA1, 0x44, 0x03, 0x42, 0x00,
+            ];
+            //        unsigned char *ptr = seckey;
+            //        memcpy(ptr, begin, sizeof(begin)); ptr += sizeof(begin);
+            //        memcpy(ptr, key32, 32); ptr += 32;
+            //        memcpy(ptr, middle, sizeof(middle)); ptr += sizeof(middle);
+            //        pubkeylen = CPubKey::PUBLIC_KEY_SIZE;
+            //        secp256k1_ec_pubkey_serialize(ctx, ptr, &pubkeylen, &pubkey, SECP256K1_EC_UNCOMPRESSED);
+            //        ptr += pubkeylen;
+            //        *seckeylen = ptr - seckey;
+            //        assert(*seckeylen == CKey::PRIVATE_KEY_SIZE);
+            SecretVec::new(
+                begin
+                    .iter()
+                    .chain(keypair.secret_bytes().iter())
+                    .chain(middle.iter())
+                    .chain(keypair.public_key().serialize_uncompressed().iter())
+                    .copied()
+                    .collect(),
+            )
+        }
+    }
+
+    // Decodes a secret key from the "openssl-inspired" DER encoding used by zcashd.
+    pub fn der_decode(encoded: &SecretVec<u8>, compressed: bool) -> Result<Self, ()> {
+        // Ported from https://github.com/zcash/zcash/blob/1f1f7a385adc048154e7f25a3a0de76f3658ca09/src/key.cpp#L36
+        // The original c++ code is retained as comments.
+
+        let seckey = encoded.expose_secret();
+
+        //    /* sequence header */
+        //    if (end - seckey < 1 || *seckey != 0x30u) {
+        //        return 0;
+        //    }
+        //    seckey++;
+        if seckey.is_empty() || seckey[0] != 0x30 {
+            return Err(());
+        }
+        let seckey = &seckey[1..];
```

Comment:
Non-blocking: Alternatives for this (and similarly below):
```suggestion
        if seckey.first() != Some(&0x30) {
            return Err(());
        }
        let seckey = &seckey[1..];
```
```suggestion
        let seckey = match seckey.split_first() {
            Some((&0x30, rest)) => Ok(rest),
            _ => Err(()),
        }?;
```